### ISSUE TITLE
mysql 5.7.5 breaks if user mounts a volume from a long running file system

### ISSUE THREAD

**Issue by jonathan-kosgei** [Percona-Lab/percona-xtradb-cluster-docker#3]
You are affected by this bug https://github.com/docker-library/mysql/issues/69, which has been driving me nuts all day. Is it possible to either downgrade your base mysql to a higher or lower stable version?

**Comment by vadimtk** [comment]
our official images are there https://hub.docker.com/r/percona/percona-xtradb-cluster/tags/
and the version we use is 5.7.17 at this moment

**Comment by jonathan-kosgei** [comment]
I'm using those and I'm trying to change the datadir mysql uses so I've made the following change:

```
	if [ -z "$DATADIR" ]; then
		DATADIR="$("mysqld" --verbose --wsrep_provider= --help 2>/dev/null | awk '$1 == "datadir" { print $2; exit }')"
	else
		echo DATADIR is set to $DATADIR
	fi
```
```
and a bit further down instead of 
mysqld --initialize-insecure
```
I have 
```
mysqld --initialize-insecure --user=mysql --datadir=$DATADIR
```

https://gist.githubusercontent.com/jonathan-kosgei/fa7dd80b0259191404177b3842d30b2c/raw/bb1b1aa1a9422fe50b56b9ee57c1b30892af1384/percona-entrypoint.sh

But I'm only able to get so far before I run into the following error; after mysqld init process done, ready for startup 

```
Registering in the discovery service

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
{"action":"create","node":{"key":"/pxc-cluster/queue/wordpress-2/00000000000000003088","value":"10.52.0.108","expiration":"2017-03-24T20:03:33.547457047Z","ttl":60,"modifiedIndex":3088,"createdIndex":3088}}
100   231  100   207  100    24  17857   2070 --:--:-- --:--:-- --:--:-- 18818
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   487  100   487    0     0  68121      0 --:--:-- --:--:-- --:--:-- 69571
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
{"action":"set","node":{"key":"/pxc-cluster/wordpress-2/10.52.0.108/ipaddr","value":"10.52.0.108","expiration":"2017-03-24T20:03:03.587150337Z","ttl":30,"modifiedIndex":3089,"createdIndex":3089}}
100   220  100   196  100    24  12826   1570 --:--:-- --:--:-- --:--:-- 13066
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
{"action":"set","node":{"key":"/pxc-cluster/wordpress-2/10.52.0.108/hostname","value":"percona-0","expiration":"2017-03-24T20:03:03.610346106Z","ttl":30,"modifiedIndex":3090,"createdIndex":3090}}
100   218  100   196  100    22  16577   1860 --:--:-- --:--:-- --:--:-- 17818
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   321  100   291  100    30  24079   2482 --:--:-- --:--:-- --:--:-- 24250
{"action":"update","node":{"key":"/pxc-cluster/wordpress-2/10.52.0.108","dir":true,"expiration":"2017-03-24T20:03:03.629957174Z","ttl":30,"modifiedIndex":3091,"createdIndex":3089},"prevNode":{"key":"/pxc-cluster/wordpress-2/10.52.0.108","dir":true,"modifiedIndex":3089,"createdIndex":3089}}
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   275  100   275    0     0  23948      0 --:--:-- --:--:-- --:--:-- 25000
Joining cluster 10.52.0.108,10.52.0.108,10.52.0.108
```

**Comment by vadimtk** [comment]
I do not see error messages here.
can you show `docker logs` from the container?

**Comment by jonathan-kosgei** [comment]
I've managed some fixes and made some progress, up until this
```
Registering in the discovery service

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:19 --:--:--     0curl: (6) Could not resolve host: etcd
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   113  100   113    0     0  14490      0 --:--:-- --:--:-- --:--:-- 16142
jq: error (at <stdin>:1): Cannot iterate over null (null)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
{"action":"set","node":{"key":"/pxc-cluster/wordpress/10.52.0.20/ipaddr","value":"10.52.0.20","expiration":"2017-03-25T20:36:51.157418283Z","ttl":30,"modifiedIndex":621,"createdIndex":621}}
100   213  100   190  100    23  11919   1442 --:--:-- --:--:-- --:--:-- 12666
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
{"action":"set","node":{"key":"/pxc-cluster/wordpress/10.52.0.20/hostname","value":"percona-0","expiration":"2017-03-25T20:36:51.181397798Z","ttl":30,"modifiedIndex":623,"createdIndex":623}}
100   213  100   191  100    22  14465   1666 --:--:-- --:--:-- --:--:-- 14692
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
{"action":"update","node":{"key":"/pxc-cluster/wordpress/10.52.0.20","dir":true,"expiration":"2017-03-25T20:36:51.201103889Z","ttl":30,"modifiedIndex":625,"createdIndex":621},"prevNode":{"key":"/pxc-cluster/wordpress/10.52.0.20","dir":true,"modifiedIndex":621,"createdIndex":621}}
100   311  100   281  100    30  21637   2310 --:--:-- --:--:-- --:--:-- 23416
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   411  100   411    0     0  28249      0 --:--:-- --:--:-- --:--:-- 29357
Joining cluster 10.52.1.31
2017-03-25T20:36:21.475363Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).
2017-03-25T20:36:21.480980Z 0 [Note] mysqld (mysqld 5.7.17-11-57) starting as process 9 ...
2017-03-25T20:36:21.492745Z 0 [Note] WSREP: Read nil XID from storage engines, skipping position init
2017-03-25T20:36:21.492803Z 0 [Note] WSREP: wsrep_load(): loading provider library '/usr/lib/galera3/libgalera_smm.so'
2017-03-25T20:36:21.496610Z 0 [Note] WSREP: wsrep_load(): Galera 3.20(r7e383f7) by Codership Oy <info@codership.com> loaded successfully.
2017-03-25T20:36:21.496770Z 0 [Note] WSREP: CRC-32C: using hardware acceleration.
2017-03-25T20:36:21.499576Z 0 [Note] WSREP: Found saved state: d79f8fa9-1199-11e7-9412-1e31718a44ec:-1, safe_to_bootsrap: 0
2017-03-25T20:36:21.506868Z 0 [Note] WSREP: Passing config to GCS: base_dir = /var/lib/mysql/percona-0/; base_host = 10.52.0.20; base_port = 4567; cert.log_conflicts = no; debug = no; evs.auto_evict = 0; evs.delay_margin = PT1S; evs.delayed_keep_period = PT30S; evs.inactive_check_period = PT0.5S; evs.inactive_timeout = PT15S; evs.join_retrans_period = PT1S; evs.max_install_timeouts = 3; evs.send_window = 4; evs.stats_report_period = PT1M; evs.suspect_timeout = PT5S; evs.user_send_window = 2; evs.view_forget_timeout = PT24H; gcache.dir = /var/lib/mysql/percona-0/; gcache.keep_pages_count = 0; gcache.keep_pages_size = 0; gcache.mem_size = 0; gcache.name = /var/lib/mysql/percona-0//galera.cache; gcache.page_size = 128M; gcache.recover = no; gcache.size = 128M; gcomm.thread_prio = ; gcs.fc_debug = 0; gcs.fc_factor = 1; gcs.fc_limit = 16; gcs.fc_master_slave = no; gcs.max_packet_size = 64500; gcs.max_throttle = 0.25; gcs.recv_q_hard_limit = 9223372036854775807; gcs.recv_q_soft_limit = 0.25; gcs.sync_donor = no; gmcast.segment = 0; gmcast.version = 0; pc.announce_timeout = PT3S; pc.checksum = false; pc.ignore_quorum = false; pc.ignore_sb = false; pc.npvo = false; pc.recovery = 1; pc.version = 0; pc.wait_prim = true; pc.wait_prim_timeout = PT30S; pc.weight = 1; protonet.backend = asio; protonet.version = 0; repl.causal_read_timeout = PT30S; repl.commit_order = 3; repl.key_format = FLAT8; repl.max_ws_size = 2147483647; repl.proto_max = 7; socket.checksum = 2; socket.recv_buf_size = 212992; 
2017-03-25T20:36:21.521501Z 0 [Note] WSREP: GCache history reset: old(d79f8fa9-1199-11e7-9412-1e31718a44ec:0) -> new(d79f8fa9-1199-11e7-9412-1e31718a44ec:-1)
2017-03-25T20:36:21.524516Z 0 [Note] WSREP: Assign initial position for certification: -1, protocol version: -1
2017-03-25T20:36:21.524647Z 0 [Note] WSREP: wsrep_sst_grab()
2017-03-25T20:36:21.524723Z 0 [Note] WSREP: Start replication
2017-03-25T20:36:21.524758Z 0 [Note] WSREP: Setting initial position to 00000000-0000-0000-0000-000000000000:-1
2017-03-25T20:36:21.525502Z 0 [Note] WSREP: protonet asio version 0
2017-03-25T20:36:21.525706Z 0 [Note] WSREP: Using CRC-32C for message checksums.
2017-03-25T20:36:21.525794Z 0 [Note] WSREP: backend: asio
2017-03-25T20:36:21.525884Z 0 [Note] WSREP: gcomm thread scheduling priority set to other:0 
2017-03-25T20:36:21.526772Z 0 [Note] WSREP: restore pc from disk successfully
2017-03-25T20:36:21.527025Z 0 [Note] WSREP: GMCast version 0
2017-03-25T20:36:21.527263Z 0 [Note] WSREP: (e54167ba, 'tcp://0.0.0.0:4567') listening at tcp://0.0.0.0:4567
2017-03-25T20:36:21.527316Z 0 [Note] WSREP: (e54167ba, 'tcp://0.0.0.0:4567') multicast: , ttl: 1
2017-03-25T20:36:21.527719Z 0 [Note] WSREP: EVS version 0
2017-03-25T20:36:21.527881Z 0 [Note] WSREP: gcomm: connecting to group 'wordpress', peer '10.52.1.31:'
2017-03-25T20:36:21.529624Z 0 [Note] WSREP: (e54167ba, 'tcp://0.0.0.0:4567') connection established to ea72b989 tcp://10.52.1.31:4567
2017-03-25T20:36:21.529784Z 0 [Note] WSREP: (e54167ba, 'tcp://0.0.0.0:4567') turning message relay requesting on, nonlive peers: 
2017-03-25T20:36:22.016396Z 0 [Note] WSREP: declaring ea72b989 at tcp://10.52.1.31:4567 stable
2017-03-25T20:36:22.017187Z 0 [Warning] WSREP: no nodes coming from prim view, prim not possible
2017-03-25T20:36:22.017256Z 0 [Note] WSREP: view(view_id(NON_PRIM,e54167ba,5) memb {
	e54167ba,0
	ea72b989,0
} joined {
} left {
} partitioned {
})
2017-03-25T20:36:22.028308Z 0 [Note] WSREP: gcomm: connected
2017-03-25T20:36:22.028475Z 0 [Note] WSREP: Changing maximum packet size to 64500, resulting msg size: 32636
2017-03-25T20:36:22.028573Z 0 [Note] WSREP: Shifting CLOSED -> OPEN (TO: 0)
2017-03-25T20:36:22.028616Z 0 [Note] WSREP: Opened channel 'wordpress'
2017-03-25T20:36:22.028726Z 0 [Note] WSREP: Waiting for SST to complete.
2017-03-25T20:36:22.029019Z 0 [Note] WSREP: New COMPONENT: primary = no, bootstrap = no, my_idx = 0, memb_num = 2
2017-03-25T20:36:22.029080Z 0 [Note] WSREP: Flow-control interval: [23, 23]
2017-03-25T20:36:22.029102Z 0 [Note] WSREP: Received NON-PRIMARY.
2017-03-25T20:36:22.029169Z 1 [Note] WSREP: New cluster view: global state: :-1, view# -1: non-Primary, number of nodes: 2, my index: 0, protocol version -1
2017-03-25T20:36:22.029217Z 1 [Note] WSREP: wsrep_notify_cmd is not defined, skipping notification.
2017-03-25T20:36:25.028945Z 0 [Note] WSREP: (e54167ba, 'tcp://0.0.0.0:4567') turning message relay requesting off
```

I get the same error on all 3 pods in a kubernetes statefulset I'm running, this is after bringing them up the first time (where they run okay) and then killing the pods and watching them restart. The pod data is backed with a volume, I was testing to see if the pods came back up with all the data intact.

**Comment by jonathan-kosgei** [comment]
Hi, I'm hesistant to open another issue so I'll post some more info:
When I start the cluster from scratch. With blank data directories and a fresh etcd cluster, everything seems to come up. However I look at the `grastate.dat` and I find that the `seq_no` for each pod is -1:

```
root@gluster-3:/mnt/gfs/gluster_vol-1/mysql# cat percona-0/grastate.dat
# GALERA saved state
version: 2.1
uuid:    a91f70f2-11f8-11e7-8f3d-86c2e58790ac
seqno:   -1
safe_to_bootstrap: 0
root@gluster-3:/mnt/gfs/gluster_vol-1/mysql# cat percona-1/grastate.dat
# GALERA saved state
version: 2.1
uuid:    a91f70f2-11f8-11e7-8f3d-86c2e58790ac
seqno:   -1
safe_to_bootstrap: 0
root@gluster-3:/mnt/gfs/gluster_vol-1/mysql# cat percona-2/grastate.dat
# GALERA saved state
version: 2.1
uuid:    a91f70f2-11f8-11e7-8f3d-86c2e58790ac
seqno:   -1
safe_to_bootstrap: 0
```
At this point I can do `mysql -h percona -u wordpress -p` and wordpress works.

Scenario 1:
I have 3 percona pods
```
/ # jonathan@ubuntu:~/Projects/k8wp$ kubectl get pods
NAME                         READY     STATUS    RESTARTS   AGE
etcd-0                       1/1       Running   1          12h
etcd-1                       1/1       Running   0          12h
etcd-2                       1/1       Running   3          12h
etcd-3                       1/1       Running   1          12h
percona-0                    1/1       Running   0          8m
percona-1                    1/1       Running   0          57m
percona-2                    1/1       Running   0          57m
```

When I try to restart `percona-0` it gets kicked out of the cluster on restarting:
percona-0's gvwstate.dat file shows
```
root@gluster-3:/mnt/gfs/gluster_vol-1/mysql# cat percona-0/gvwstate.dat
my_uuid: b7571ff8-11f8-11e7-bd2d-8b50487e1523
#vwbeg
view_id: 3 b7571ff8-11f8-11e7-bd2d-8b50487e1523 3
bootstrap: 0
member: b7571ff8-11f8-11e7-bd2d-8b50487e1523 0
member: bd05a643-11f8-11e7-9dab-1b4fc20eaf6a 0
member: c33d6a73-11f8-11e7-9e86-fe1cf3d3367a 0
#vwend
```

The other 2 pods in the cluster show:
```
root@gluster-3:/mnt/gfs/gluster_vol-1/mysql# cat percona-1/gvwstate.dat
my_uuid: bd05a643-11f8-11e7-9dab-1b4fc20eaf6a
#vwbeg
view_id: 3 bd05a643-11f8-11e7-9dab-1b4fc20eaf6a 4
bootstrap: 0
member: bd05a643-11f8-11e7-9dab-1b4fc20eaf6a 0
member: c33d6a73-11f8-11e7-9e86-fe1cf3d3367a 0
#vwend
root@gluster-3:/mnt/gfs/gluster_vol-1/mysql# cat percona-2/gvwstate.dat
my_uuid: c33d6a73-11f8-11e7-9e86-fe1cf3d3367a
#vwbeg
view_id: 3 bd05a643-11f8-11e7-9dab-1b4fc20eaf6a 4
bootstrap: 0
member: bd05a643-11f8-11e7-9dab-1b4fc20eaf6a 0
member: c33d6a73-11f8-11e7-9e86-fe1cf3d3367a 0
#vwend
```

Here are what I think are the relevant errors from `percona-0's` startup:
```
2017-03-26T08:37:58.370605Z 0 [Note] WSREP: Setting initial position to 00000000-0000-0000-0000-000000000000:-1
```
```
2017-03-26T08:37:58.372537Z 0 [Note] WSREP: gcomm: connecting to group 'wordpress-001', peer '10.52.0.26:'
2017-03-26T08:38:01.373345Z 0 [Note] WSREP: (b7571ff8, 'tcp://0.0.0.0:4567') connection to peer 00000000 with addr tcp://10.52.0.26:4567 timed out, no messages seen in PT3S
2017-03-26T08:38:01.373682Z 0 [Warning] WSREP: no nodes coming from prim view, prim not possible
2017-03-26T08:38:01.373750Z 0 [Note] WSREP: view(view_id(NON_PRIM,b7571ff8,5) memb {
	b7571ff8,0
} joined {
} left {
} partitioned {
})
2017-03-26T08:38:01.373838Z 0 [Note] WSREP: gcomm: connected
2017-03-26T08:38:01.373872Z 0 [Note] WSREP: Changing maximum packet size to 64500, resulting msg size: 32636
2017-03-26T08:38:01.373987Z 0 [Note] WSREP: Shifting CLOSED -> OPEN (TO: 0)
2017-03-26T08:38:01.374012Z 0 [Note] WSREP: Opened channel 'wordpress-001'
2017-03-26T08:38:01.374108Z 0 [Note] WSREP: Waiting for SST to complete.
2017-03-26T08:38:01.374417Z 0 [Note] WSREP: New COMPONENT: primary = no, bootstrap = no, my_idx = 0, memb_num = 1
2017-03-26T08:38:01.374469Z 0 [Note] WSREP: Flow-control interval: [16, 16]
2017-03-26T08:38:01.374491Z 0 [Note] WSREP: Received NON-PRIMARY.
2017-03-26T08:38:01.374560Z 1 [Note] WSREP: New cluster view: global state: :-1, view# -1: non-Primary, number of nodes: 1, my index: 0, protocol version -1
```

The ip it's trying to connect to `10.52.0.26` in `2017-03-26T08:37:58.372537Z 0 [Note] WSREP: gcomm: connecting to group 'wordpress-001', peer '10.52.0.26:'` is actually that pods previous ip, here's the listing of keys in etcd I did before deleting `percona-0`

```
/ # etcdctl ls --recursive
/pxc-cluster
/pxc-cluster/wordpress
/pxc-cluster/queue
/pxc-cluster/queue/wordpress
/pxc-cluster/queue/wordpress-001
/pxc-cluster/wordpress-001
/pxc-cluster/wordpress-001/10.52.1.46
/pxc-cluster/wordpress-001/10.52.1.46/ipaddr
/pxc-cluster/wordpress-001/10.52.1.46/hostname
/pxc-cluster/wordpress-001/10.52.2.33
/pxc-cluster/wordpress-001/10.52.2.33/ipaddr
/pxc-cluster/wordpress-001/10.52.2.33/hostname
/pxc-cluster/wordpress-001/10.52.0.26
/pxc-cluster/wordpress-001/10.52.0.26/hostname
/pxc-cluster/wordpress-001/10.52.0.26/ipaddr
```
After `kubectl delete pods/percona-0`:
```
/ # etcdctl ls --recursive
/pxc-cluster
/pxc-cluster/queue
/pxc-cluster/queue/wordpress
/pxc-cluster/queue/wordpress-001
/pxc-cluster/wordpress-001
/pxc-cluster/wordpress-001/10.52.1.46
/pxc-cluster/wordpress-001/10.52.1.46/ipaddr
/pxc-cluster/wordpress-001/10.52.1.46/hostname
/pxc-cluster/wordpress-001/10.52.2.33
/pxc-cluster/wordpress-001/10.52.2.33/ipaddr
/pxc-cluster/wordpress-001/10.52.2.33/hostname
/pxc-cluster/wordpress
```
Also during the restart `percona-0` tried to register to etcd with:
```
{"action":"create","node":{"key":"/pxc-cluster/queue/wordpress-001/00000000000000009886","value":"10.52.0.27","expiration":"2017-03-26T08:38:57.980325718Z","ttl":60,"modifiedIndex":9886,"createdIndex":9886}}
{"action":"set","node":{"key":"/pxc-cluster/wordpress-001/10.52.0.27/ipaddr","value":"10.52.0.27","expiration":"2017-03-26T08:38:28.01814818Z","ttl":30,"modifiedIndex":9887,"createdIndex":9887}}
{"action":"set","node":{"key":"/pxc-cluster/wordpress-001/10.52.0.27/hostname","value":"percona-0","expiration":"2017-03-26T08:38:28.037188157Z","ttl":30,"modifiedIndex":9888,"createdIndex":9888}}
{"action":"update","node":{"key":"/pxc-cluster/wordpress-001/10.52.0.27","dir":true,"expiration":"2017-03-26T08:38:28.054726795Z","ttl":30,"modifiedIndex":9889,"createdIndex":9887},"prevNode":{"key":"/pxc-cluster/wordpress-001/10.52.0.27","dir":true,"modifiedIndex":9887,"createdIndex":9887}}
```
which doesn't work.

So in summary when starting a fresh percona cluster of 3 kubernetes pods, the grastate.dat seq_no get's stuck at -1. On deleting one pod and watching it restart, expecting it to rejoin the cluster, it sets it's inital position to `00000000-0000-0000-0000-000000000000:-1` and tries to connect to itself (it's former ip), maybe because it'd been the first pod in the cluster? It then timeouts in it's erroneous connection to itself:
```
2017-03-26T08:38:05.374058Z 0 [Note] WSREP: (b7571ff8, 'tcp://0.0.0.0:4567') connection to peer 00000000 with addr tcp://10.52.0.26:4567 timed out, no messages seen in PT3S
```
The cluster doesn't get started properly and I'm unable to restart the cluster.

**Comment by jonathan-kosgei** [comment]
Fixed it with changing the entrypoint in the container to the following script:

```
#!/bin/bash
sed -i \"s|safe_to_bootstrap.*:.*|safe_to_bootstrap:1|1\" /var/lib/mysql/`hostname`/grastate.dat; 
/entrypoint.sh --wsrep-new-cluster;
```
Thanks. Ref -> https://www.claudiokuenzler.com/blog/494/galera-cluster-mysql-not-starting-failed-to-open-channel-reach-primary#.WNesDiF97Qo
